execution/tvf/percentile_watermark_generator.go

```diff
@@ -0,0 +1,148 @@
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source")
+	}
+
+	events, err := w.events.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark events")
+	}
+	if events.GetType() != octosql.TypeDuration || events.AsInt() < 1 {
+		return nil, nil, errors.Errorf("invalid watermark events: %v", events)
+	}
+
+	percentile, err := w.percentile.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark percentile")
+	}
+	if percentile.GetType() != octosql.TypeDuration || percentile.AsInt() < 1 || percentile.AsInt() > 99 {
+		return nil, nil, errors.Errorf("invalid watermark percentile: %v", percentile)
+	}
+
+	ws := &PercentileWatermarkGeneratorStream{
+		source:     source,
+		timeField:  w.timeField,
+		events:     events.AsInt(),
+		percentile: percentile.AsInt(),
+		eventsSeen: 0,
+	}
+
+	return ws, execution.NewExecOutput(ws), nil // watermark generator stream now indicates new watermark source
+}
+
+type PercentileWatermarkGeneratorStream struct {
+	source     execution.RecordStream
+	timeField  octosql.VariableName
+	events     int
+	percentile int
+	eventsSeen int
+}
+
+var percentileWatermarkPrefix = []byte("$percentile_watermark$")
+var percentileWatermarkEventsPrefix = []byte("$percentile_watermark_events$")
+
+func (s *PercentileWatermarkGeneratorStream) GetWatermark(ctx context.Context, tx storage.StateTransaction) (time.Time, error) {
+	watermarkStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkPrefix))
+
+	var currentWatermark octosql.Value
+	err := watermarkStorage.Get(&currentWatermark)
+	if err == storage.ErrNotFound {
+		currentWatermark = octosql.MakeTime(time.Time{})
+	} else if err != nil {
+		return time.Time{}, errors.Wrap(err, "couldn't get current watermark from storage")
+	}
+
+	return currentWatermark.AsTime(), nil
+}
+
+func (s *PercentileWatermarkGeneratorStream) Next(ctx context.Context) (*execution.Record, error) {
+	srcRecord, err := s.source.Next(ctx)
+	if err != nil {
+		if err == execution.ErrEndOfStream {
+			return nil, execution.ErrEndOfStream
+		}
+		return nil, errors.Wrap(err, "couldn't get source record")
+	}
+
+	timeValue := srcRecord.Value(s.timeField)
+	if timeValue.GetType() != octosql.TypeTime {
+		return nil, fmt.Errorf("couldn't get time field '%v' as time, got: %v", s.timeField.String(), srcRecord.Value(s.timeField))
+	}
+
+	tx := storage.GetStateTransactionFromContext(ctx)
+	watermarkStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkPrefix))
+	eventsStorage := storage.NewDeque(tx.WithPrefix(percentileWatermarkEventsPrefix))
+
+	var watermarkPlace int       // represents position of event in sorted list that will become new watermark
+	if s.eventsSeen < s.events { // no need to pop first element from deque
+		s.eventsSeen++
```

Comment:
Not sure that's the case here, but shouldn't `eventsSeen` be kept on disk in case of machine failure, so that the generator can restore itself correctly?

Reply:
You could just use the deque length I think?

Reply:
There is no method to extract deque length

Reply:
Add it then.